modules/tinymce/src/core/test/ts/browser/lookup/UserLookupTest.ts

```diff
@@ -0,0 +1,345 @@
+      id: userId,
+      name: 'Test User',
+      avatar: 'test-avatar.png',
+      description: 'Test Description'
+    };
+
+    await expect(userPromise).to.eventually.deep.equal(
+      expectedUser,
+      'Should resolve with expected user object'
+    );
+  });
+
+  it('TINY-11974: Should cache and return same data for subsequent requests', async () => {
+    const editor = hook.editor();
+    const userId = 'test-user-1';
+
+    const [ firstPromise ] = editor.userLookup.fetchUsers([ userId ]);
+    const [ secondPromise ] = editor.userLookup.fetchUsers([ userId ]);
+
+    await expect(firstPromise).to.eventually.deep.equal(
+      await secondPromise,
+      'Should return same data for subsequent requests'
+    );
+  });
+
+  it('TINY-11974: Should reject promise for non-existent users', async () => {
+    const editor = hook.editor();
+    const nonExistentId = 'non-existent-user';
+
+    const [ userPromise ] = editor.userLookup.fetchUsers([ nonExistentId ]);
+
+    await expect(userPromise).to.eventually.be.rejectedWith(
+      Error,
+      `User ${nonExistentId} not found`,
+      'Should reject for non-existent user'
+    );
+  });
+
+  it('TINY-11974: Should handle repeated requests for non-existent users', async () => {
+    const editor = hook.editor();
+    const nonExistentId = 'non-existent-user';
+    const errorMessage = `User ${nonExistentId} not found`;
+
+    const [ firstPromise ] = editor.userLookup.fetchUsers([ nonExistentId ]);
+    await expect(firstPromise).to.eventually.be.rejectedWith(errorMessage);
+
+    // Second request should use cached rejection
+    const [ secondPromise ] = editor.userLookup.fetchUsers([ nonExistentId ]);
+    await expect(secondPromise).to.eventually.be.rejectedWith(errorMessage);
+  });
+
+  it('TINY-11974: Should handle mix of existing and non-existent users', async () => {
+    const editor = hook.editor();
+    const existingId = 'test-user-1';
+    const nonExistentId = 'non-existent-user';
+
+    const [ existingPromise, nonExistentPromise ] = editor.userLookup.fetchUsers([ existingId, nonExistentId ]);
+
+    await expect(existingPromise).to.eventually.deep.equal({
+      id: existingId,
+      name: 'Test User',
+      avatar: 'test-avatar.png',
+      description: 'Test Description'
+    });
+
+    await expect(nonExistentPromise).to.be.rejectedWith(`User ${nonExistentId} not found`);
+  });
+
+  it('TINY-11974: Should handle concurrent requests for the same user', async () => {
+    const editor = hook.editor();
+    const userId = 'test-user-1';
+
+    // Make multiple concurrent requests
+    const [ promise1, promise2, promise3 ] = [
+      editor.userLookup.fetchUsers([ userId ])[0],
+      editor.userLookup.fetchUsers([ userId ])[0],
+      editor.userLookup.fetchUsers([ userId ])[0]
+    ];
+
+    await Promise.all([
+      expect(promise1).to.eventually.deep.equal(await promise2),
+      expect(promise2).to.eventually.deep.equal(await promise3),
+    ]);
```

Comment:
the nested `await` means the promises are already resolved before you call `Promise.all`. And you don't need to `all` this anyway.


```suggestion
    // the values are asserted in other tests, so just assert they are the same
    await expect(promise1).to.eventually.deep.equal(promise2);
    await expect(promise2).to.eventually.deep.equal(promise3);
```

Reply:
I think you meant:

```ts
    await expect(promise1).to.eventually.deep.equal(await promise2);
    await expect(promise2).to.eventually.deep.equal(await promise3);
```

Since in the code block you've shared, `eventually` tells chai to wait for `promise1` to resolve whereas `promise2` has been passed in as unresolved so we are essentially comparing a resolved promise with an unresolved promise. Same thing with the next line.

Reply:
I specifically want to avoid the `await promise2` inside the deep equal. When you do that, the later `await expect(promise2)` is a bit pointless because it's already been `await`ed (although, to be fair, that's going to be true either way).

It looks like chai as expected might offer that as `expect(promise1).becomes(promise2)`?

Or perhaps avoid all the challenging asynchronicity, do a `Promise.all([1, 2, 3])` and synchronous assertions on the values.

Reply:
I went with the latter approach by using `Promise.all` and then handling each one synchronously.